adapters/bigoad/bigoad.go

```diff
@@ -0,0 +1,163 @@
+		"Content-Type":      "application/json;charset=utf-8",
+		"Accept":            "application/json",
+		"x-openrtb-version": "2.5",
+	})
+	return headers
+}
+
+func addNonEmptyHeaders(headers *http.Header, headerValues map[string]string) {
+	for key, value := range headerValues {
+		if len(value) > 0 {
+			headers.Add(key, value)
+		}
+	}
+}
+
+func getImpExt(imp *openrtb2.Imp) (*openrtb_ext.ExtImpBigoAd, error) {
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: fmt.Sprintf("imp %s: unable to unmarshal ext", imp.ID),
+		}
+	}
+	var bigoadExt openrtb_ext.ExtImpBigoAd
+	if err := json.Unmarshal(bidderExt.Bidder, &bigoadExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: fmt.Sprintf("imp %s: unable to unmarshal ext.bidder: %v", imp.ID, err),
+		}
+	}
+	imp.Ext = bidderExt.Bidder
+	return &bigoadExt, nil
+}
+
+func (a *adapter) buildEndpointURL(params *openrtb_ext.ExtImpBigoAd) (string, error) {
+	endpointParams := macros.EndpointTemplateParams{Host: params.Host, SspId: params.SspId}
+	return macros.ResolveMacros(a.endpoint, endpointParams)
+}
+
+func (a *adapter) MakeBids(request *openrtb2.BidRequest, _ *adapters.RequestData, responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if adapters.IsResponseStatusCodeNoContent(responseData) {
+		return nil, nil
+	}
+
+	if err := adapters.CheckResponseStatusCodeForErrors(responseData); err != nil {
+		return nil, []error{err}
+	}
+
+	var bidResponse openrtb2.BidResponse
+	if err := json.Unmarshal(responseData.Body, &bidResponse); err != nil {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Bad server response: %d", err),
+		}}
+	}
+
+	if len(bidResponse.SeatBid) == 0 {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: "Empty SeatBid array",
+		}}
+	}
+
+	bidResponseWithCapacity := adapters.NewBidderResponseWithBidsCapacity(len(bidResponse.SeatBid[0].Bid))
+
+	var errors []error
+	seatBid := bidResponse.SeatBid[0]
+	for i := range seatBid.Bid {
+		bid := seatBid.Bid[i]
+		bidType, err := getBidType(request.Imp[0], bid)
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+		bidResponseWithCapacity.Bids = append(bidResponseWithCapacity.Bids, &adapters.TypedBid{
+			Bid:     &bid,
+			BidType: bidType,
+		})
+	}
+
+	return bidResponseWithCapacity, errors
+}
+
+func getBidType(imp openrtb2.Imp, bid openrtb2.Bid) (openrtb_ext.BidType, error) {
+	switch bid.MType {
+	case openrtb2.MarkupBanner:
+		return openrtb_ext.BidTypeBanner, nil
+	case openrtb2.MarkupNative:
+		return openrtb_ext.BidTypeNative, nil
+	case openrtb2.MarkupVideo:
+		return openrtb_ext.BidTypeVideo, nil
+	}
+	if imp.Native != nil {
+		return openrtb_ext.BidTypeNative, nil
+	} else if imp.Banner != nil {
+		return openrtb_ext.BidTypeBanner, nil
+	} else if imp.Video != nil {
+		return openrtb_ext.BidTypeVideo, nil
+	}
+
```

Comment:
@BIGOAds any particular reason to fallback to above logic when no match is found for MType?